internal/fans/nvidia.go

```diff
@@ -0,0 +1,228 @@
+	if fan.StartPwm != nil {
+		return *fan.StartPwm
+	} else {
+		return 100
+	}
+}
+
+func (fan *NvidiaFan) SetStartPwm(pwm int, force bool) {
+	if fan.Config.StartPwm == nil || force {
+		fan.StartPwm = &pwm
+	}
+}
+
+func (fan *NvidiaFan) GetMaxPwm() int {
+	if fan.MaxPwm != nil {
+		return *fan.MaxPwm
+	} else {
+		return 100
+	}
+}
+
+func (fan *NvidiaFan) SetMaxPwm(pwm int, force bool) {
+	if fan.Config.MaxPwm == nil || force {
+		fan.MaxPwm = &pwm
+	}
+}
+
+func (fan *NvidiaFan) GetRpm() (int, error) {
+	return fan.GetPwm()
+	// TODO: for driver 565 and newer we can get the actual RPM with custom cgo
+	// code, see my_DeviceGetFanSpeedRPM() in Daniel's nvml test code.
+	// But otherwise, I guess returning the current PWM should also work as it
+	// at least allows creating a fan curve
+}
+
+func (fan *NvidiaFan) GetRpmAvg() float64 {
+	return fan.RpmMovingAvg
+}
+
+func (fan *NvidiaFan) SetRpmAvg(rpm float64) {
+	fan.RpmMovingAvg = rpm
+}
+
+func (fan *NvidiaFan) GetPwm() (int, error) {
+	var device nvml.Device = nil // TODO!
+	fanIdx := fan.getNvFanIndex()
+	speed, ret := device.GetFanSpeed_v2(fanIdx)
+	if ret != nvml.SUCCESS {
+		speed = MinPwmValue // this is what HwMonFan does
+	}
+	// TODO: convert speed from percent (0..100) to PWM (0..255)?
+	return int(speed), nvError(ret)
+}
+
+func (fan *NvidiaFan) SetPwm(pwm int) (err error) {
+	ui.Debug("Setting Fan PWM of '%s' to %d ...", fan.GetId(), pwm)
+	var device nvml.Device = nil // TODO!
+	fanIdx := fan.getNvFanIndex()
+	// TODO: translate pwm (0..255) to percent (0..100)?
+	// or just clamp to 100 and let fan2go assume that it doesn't get faster after PWM > 100?
+	pwm = min(pwm, 100)
+	ret := device.SetFanSpeed_v2(fanIdx, pwm)
+	return nvError(ret)
+}
+
+func (fan *NvidiaFan) GetFanRpmCurveData() *map[int]float64 {
+	return fan.FanCurveData
+}
+
+// AttachFanCurveData attaches fan curve data from persistence to a fan
+// Note: When the given data is incomplete, all values up until the highest
+// value in the given dataset will be interpolated linearly
+// returns os.ErrInvalid if curveData is void of any data
+func (fan *NvidiaFan) AttachFanRpmCurveData(curveData *map[int]float64) (err error) {
+	if curveData == nil || len(*curveData) <= 0 {
+		ui.Error("Cant attach empty fan curve data to fan %s", fan.GetId())
+		return os.ErrInvalid
+	}
+
+	fan.FanCurveData = curveData
+
+	startPwm, maxPwm := ComputePwmBoundaries(fan)
+	fan.SetStartPwm(startPwm, false)
+	fan.SetMaxPwm(maxPwm, false)
+
+	// TODO: we don't have a way to determine this yet
+	fan.SetMinPwm(startPwm, false)
+
+	return err
+}
+
+func (fan *NvidiaFan) UpdateFanRpmCurveValue(pwm int, rpm float64) {
+	if fan.FanCurveData == nil {
+		fan.FanCurveData = &map[int]float64{}
+	}
+	(*fan.FanCurveData)[pwm] = rpm
+}
+
+func (fan *NvidiaFan) GetCurveId() string {
+	return fan.Config.Curve
+}
+
+func (fan *NvidiaFan) ShouldNeverStop() bool {
+	return fan.Config.NeverStop
+}
+
+func (fan *NvidiaFan) GetPwmEnabled() (int, error) {
```

Comment:
I think we need to change the signature of this method.
The "pwm enabled" values are a result of hwmon using it to indicate the mode. What we actually need is an abstraction on top of that, which is specified inside of fan2go and agnostic of the used platform. That way f.ex. a `file` fan could also specify this mode using `manual`, `auto`, `max` instead if `0`, `1` and `2`.

Reply:
Since your code works fine as is, I think I can do this in a separate PR.
I will create a new issue for this.